File: solutions/1-two-sum.js
// Given an array of integers nums and an integer target, return indices
// of the two numbers such that they add up to target

// You may assume that each input would have exactly one solution, and you
// may not use the same element twice

// O(n^2)
// function solution(nums, target) {
//   let left = 0;
//   let right = 1;
//   let sum;

//   while (left < nums.length + 1) {
//     sum = nums[left] + nums[right];
//     if (sum === target) {
//       return [left, right];
//     } else if (right < nums.length + 1) {
//       right++;
//     } else {
//       left++;
//       right = left + 1;
//     }
//   }
// }

// O(n)
function solution(nums, target) {
  let map = new Map();

  for (i = 0; i < nums.length; i++) {
    if (map.has(target - nums[i])) {
      return [map.get(target - nums[i]), i];
    } else {
      map.set(nums[i], i);
    }
  }
}

console.log("Result: [" + solution([2, 7, 11, 15], 9) + "] Solution: [0,1]");
console.log("Result: [" + solution([3, 2, 4], 6) + "] Solution: [1,2]");
console.log("Result: [" + solution([3, 3], 6) + "] Solution: [0,1]");

File: c#/c#/1-TwoSum.cs
class TwoSum
{
    static int[] Solution(int[] nums, int target)
    {
        //int[] result = new int[2];
        //for (int i = 0; i < nums.Length; i++)
        //{
        //    for (int j = i + 1; j < nums.Length; j++)
        //    {
        //        if (nums[i] + nums[j] == target) 
        //        {
        //            result[0] = i;
        //            result[1] = j;
        //            break;
        //        }
        //    }
        //}
        //return result;

        Dictionary<int, int> dict = new Dictionary<int, int>
    {
        { nums[0], 0 }
    };
        for (int i = 1; i < nums.Length; i++)
        {
            if (dict.ContainsKey(target - nums[i]))
            {
                return new int[] { dict[target - nums[i]], i };
            }
            else if (!dict.ContainsKey(nums[i]))
            {
                dict.Add(nums[i], i);
            }
        }
        return new int[] { };
    }

}